lxd/api_project.go

```diff
@@ -1078,6 +1080,252 @@ func projectNodeConfigDelete(d *Daemon, s *state.State, name string) error {
 	return nil
 }
 
+// doProjectForceDelete handles force deletion of project entities.
+func doProjectForceDelete(ctx context.Context, s *state.State, projectName string) error {
+	// Get the project entities and config.
+	var usedBy []string
+	var projectConfig map[string]string
+	err := s.DB.Cluster.Transaction(ctx, func(ctx context.Context, tx *db.ClusterTx) error {
+		project, err := dbCluster.GetProject(ctx, tx.Tx(), projectName)
+		if err != nil {
+			return fmt.Errorf("Fetch project %q: %w", projectName, err)
+		}
+
+		usedBy, err = projectUsedBy(ctx, tx, project)
+		if err != nil {
+			return err
+		}
+
+		projectConfig, err = dbCluster.GetProjectConfig(ctx, tx.Tx(), project.Name)
+		if err != nil {
+			return fmt.Errorf("Fetch project %q config: %w", projectName, err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Parse and delete entities.
+	defaultProfile := api.NewURL().Path(version.APIVersion, "profiles", api.ProjectDefaultName).Project(projectName).String()
+	localNodeName := s.ServerName
+
+	for _, u := range usedBy {
+		if u == defaultProfile {
+			continue
+		}
+
+		url, err := url.Parse(u)
+		if err != nil {
+			return err
+		}
+
+		entityType, _, target, pathArgs, err := entity.ParseURL(*url)
+		if err != nil {
+			return err
+		}
+
+		// Skip entities that are not local to this node (only applies to instances and storage volumes).
+		if (entityType == entity.TypeInstance || entityType == entity.TypeStorageVolume) && target != "" && target != localNodeName {
+			continue
+		}
+
+		// Delete the entity based on its type.
+		switch entityType {
+		case entity.TypeInstance:
+			instanceName := pathArgs[0]
+			err = s.Authorizer.CheckPermission(ctx, entity.InstanceURL(projectName, instanceName), auth.EntitlementCanDelete)
```

Comment:
@markylaing if you have can delete on a project, do you implicitly have can delete on all entities inside that project, or do we need to check each entity separately?

Reply:
You need to check everything separately.

@kadinsayani we'll need to do this first when parsing URLs before actually deleting anything so that the whole action either succeeds or fails